src/hotspot/share/interpreter/oopMapCache.cpp

```diff
@@ -500,8 +501,10 @@ void OopMapCache::flush_obsolete_entries() {
           ("flush: %s(%s): cached entry @%d",
            entry->method()->name()->as_C_string(), entry->method()->signature()->as_C_string(), i);
       }
-      _array[i] = nullptr;
-      OopMapCacheEntry::deallocate(entry);
+
+      if (Atomic::cmpxchg(&_array[i], entry, (OopMapCacheEntry*)nullptr, memory_order_relaxed) == entry) {
+        enqueue_for_cleanup(entry);
```

Comment:
OopMapCache::flush_obsolete_entries() is called from VM_RedefineClasses::doit(). Enqueuing  OopMapCacheEntries for cleanup means that we would need call OopMapCache::cleanup_old_entries() in VM_RedefineClasses::doit_epilogue()

Reply:
Good catch. Fixed.